### PR TITLE
feat: Use user ID to track the rate limit

### PR DESCRIPTION
**Pull request by davidmytton**
Add the user ID to the session and use it for tracking the rate limit so that if the limit is hit, you can sign in and get a new limit linked to your session.

### CODE REVIEW

**Comment by blaine-arcjet** [comment]
> you can sign in and get a new limit linked to your session.

I believe you will still be in the block cache. As we've seen in [ben's example](https://github.com/arcjet/arcjet-js/blob/4dac6d5c1f515960b54d215218cdeb5e3030c3f5/examples/nextjs-14-permit/src/lib/arcjet.ts#L8-L9), we need to use something like `fingerprint` as the top-level characteristic and vary it based on login vs ip.

**Comment by davidmytton** [comment]
When I tested it the block cache seemed to be purged when I logged out, but I've switched it to the method in Ben's example anyway.

File: app/rate-limiting/test/route.ts [hunk 1]
```diff
@@ -13,24 +13,29 @@ const aj = arcjet.withRule(
...
 // Returns ad-hoc rules depending on whether the session is present. You could
 // inspect more details about the session to dynamically adjust the rate limit.
 function getClient(session: Session | null) {
-  // If the user is logged in then give them a higher rate limit
+  // If the user is logged in then give them a higher rate limit and track the
+  // limit based on their user ID
   if (session?.user) {
     return aj.withRule(
       fixedWindow({
+        characteristics: ["userId"],
```
Comment (blaine-arcjet) [inline app/rate-limiting/test/route.ts:27]:
```suggestion
```
You'll just want to use the fingerprint

File: app/rate-limiting/test/route.ts [hunk 2]
```diff
@@ -13,24 +13,29 @@ const aj = arcjet.withRule(
...
-  })
+  }),
 );
 
 // Returns ad-hoc rules depending on whether the session is present. You could
 // inspect more details about the session to dynamically adjust the rate limit.
 function getClient(session: Session | null) {
-  // If the user is logged in then give them a higher rate limit
+  // If the user is logged in then give them a higher rate limit and track the
+  // limit based on their user ID
```
Comment (blaine-arcjet) [inline app/rate-limiting/test/route.ts:23]:
```suggestion
```

File: app/rate-limiting/test/route.ts [hunk 3]
```diff
@@ -13,24 +13,29 @@ const aj = arcjet.withRule(
...
       fixedWindow({
+        characteristics: ["userId"],
         mode: "LIVE",
         max: 5,
         window: "60s",
       }),
     );
   } else {
+    // If the user is not logged in then give them a lower rate limit and use
+    // the IP address to track them
```
Comment (blaine-arcjet) [inline app/rate-limiting/test/route.ts:35]:
```suggestion
```

File: app/rate-limiting/test/route.ts [hunk 4]
```diff
@@ -13,24 +13,29 @@ const aj = arcjet.withRule(
...
     return aj.withRule(
       fixedWindow({
+        characteristics: ["ip.src"],
```
Comment (blaine-arcjet) [inline app/rate-limiting/test/route.ts:38]:
```suggestion
```

File: app/rate-limiting/test/route.ts [hunk 5]
```diff
@@ -39,15 +44,21 @@ function getClient(session: Session | null) {
...
 
+  // Use the user ID if the user is logged in, otherwise use the IP address
+  const fingerprint = session?.user?.id ?? req.ip!;
+
   // The protect method returns a decision object that contains information
   // about the request.
-  const decision = await getClient(session).protect(req);
+  const decision = await getClient(session).protect(req, {
+    fingerprint,
+    userId: session?.user?.id!,
```
Comment (blaine-arcjet) [inline app/rate-limiting/test/route.ts:60]:
```suggestion
```

**Review by blaine-arcjet — CHANGES_REQUESTED** [review]
We want to remove the characteristics from the rate limit rules and just use the fingerprint as defined as either the `userId` or `ip`.

**Review by blaine-arcjet — APPROVED** [review]

**Pull request merged**